### DISCUSSION TITLE
How to use `PCHPDDM` in petsc+hpddm?

### DISCUSSION THREAD

**Discussion by lrtfm** [spack/spack#36082] [General]
It seems that `PCHPDDM` needs `petsc` built with `slepc`. Is there a way to enable `PCHPDDM` in `petsc` built by spack? 

@balay  @barrysmith  @jedbrown

Thank you for your attention.

**Comment by balay** [comment]
Hm - I see some support was added in https://github.com/spack/spack/pull/24725 -  but I'm not sure if its possible to manage this circular dependency on slepc in spack.. 

cc: @prj-

**Reply by prj-** [reply]
You can enable `PCHPDDM` by `slepc` built by spack if your PETSc does not come with `--download-slepc --download-hpddm`. There is not standalone support for `PCHPDDM` without `--download-hpddm` in PETSc (+ `--download-slepc`) or in SLEPc, but that is something I'd like to address (though it would need to tinker around with `LD_PRELOAD`, I think, so not ideal).
Just to sum up, as of today: either PETSc with `--download-slepc` and `--download-hpddm`, or PETSc with neither `--download-slepc` nor `--download-hpddm` and SLEPc with `--download-hpddm`.

**Reply by prj-** [reply]
Not my finest work, but here is a proof of concept: https://gitlab.com/petsc/petsc/-/merge_requests/6181. With this branch, you can:
* compile PETSc with `--download-hpddm` but _without_ `--download-slepc`. Only `KSPHPDDM` will be available at this point in time, if you try to use `PCHPDDM` you'll get a PETSc error;
* compile SLEPc _without_ `--download-hpddm` (still no `PCHPDDM` at this point in time, only `KSPHPDDM`);
* compile `libhpddm_petsc.sharedLibraryExt` and then `PCHPDDM` will be available as well.

For the first step, you must define `PETSC_DIR` and `PETSC_ARCH`, for the second step, you must define `SLEPC_DIR`, for the third step, you must define `HPDDM_DIR`. I have yet to try with prefix installs (which is what is done in spack if I were to guess?).

**Reply by lrtfm** [reply]
> Just to sum up, as of today: either PETSc with `--download-slepc` and `--download-hpddm`, or PETSc with neither `--download-slepc` nor `--download-hpddm` and SLEPc with `--download-hpddm`.

Do these two methods give the same result, i.e. `KSPHPDDM` and `PCHPDDM` all enabled? If this is true, adding HPDDM to SLEPC seems a good choice in spack.

**Reply by lrtfm** [reply]
> I have yet to try with prefix installs (which is what is done in spack if I were to guess?).
Yes, spack install packages all with prefix. Then the third step could be done as a `post_install` action of SLEPC in spack, I guess.

**Reply by prj-** [reply]
> Do these two methods give the same result, i.e. KSPHPDDM and PCHPDDM all enabled?

Yes.

>  If this is true, adding HPDDM to SLEPC seems a good choice in spack.

Yes, I'm working on that.

**Reply by prj-** [reply]
The two MR to keep track of are: [PETSc one](https://gitlab.com/petsc/petsc/-/merge_requests/6181) and [SLEPc one](https://gitlab.com/slepc/slepc/-/merge_requests/505). I'm not proficient enough with spack, but once these MR are cleaned up and merged, you should be able to `spack install petsc +hpddm` (only `KSPHPDDM` available at this point in time) and then `spack install slepc +hpddm` (both `KSPHPDDM` and `PCHPDDM` available). I'll have to reach out to @balay or @joseeroman to help me update the SLEPc spack `package.py`.

**Reply by prj-** [reply]
Here is the [spack PR](https://github.com/spack/spack/pull/36118).

**Reply by lrtfm** [reply]
Hi @prj, thank you for your prompt work. I have a quick question regarding the dependency issue with HPDDM. Could you please clarify why it's not working as a dependency in space? Is it because we need the source code of HPDDM?

**Reply by prj-** [reply]
> Could you please clarify why it's not working as a dependency in space?

I don't understand this question.

**Reply by prj-** [reply]
Maybe you meant `dependency in spack`? Still, I don't really understand the question. One thing for sure, you should not use the `hpddm` spack package: it's outdated (and unmaintained?) and only good if you are interested in HPDDM _without_ PETSc using a native interface. So it's no good if you want to use `PCHPDDM` in PETSc.

**Comment by lrtfm** [comment]
> Maybe you meant dependency in spack? Still, I don't really understand the question. 

Yes, I mean the dependency in spack. petsc and slepc now download the package hpddm other than using the one provided by spack. 

> One thing for sure, you should not use the hpddm spack package: it's outdated (and unmaintained?) and only good if you are interested in HPDDM without PETSc using a native interface. So it's no good if you want to use PCHPDDM in PETSc.

Thanks for your clarifications.

**Comment by lrtfm** [comment]
I need to install packages, such as `petsc` and `slepc`, on a system without internet access using spack's mirror function. However, when using the `--download-hpddm` option, I encounter an error due to the host's inability to access the internet.

To address this issue, I added new versions of `hpddm` to spack (#36769) and attempted to use the option `--with-hpddm-dir` (or `--with-hpddm-include`, `--with-hpddm-lib`), but found that it is not supported. Is there a way to make the option `--with-hpddm-dir` (or `--with-hpddm-include`, `--with-hpddm-lib`) work?"

@prj - Would you mind taking a look at this, please?

**Reply by prj-** [reply]
We could make it work, but it's touchy because HPDDM is a weird PETSc package (`self.builtafterpetsc = 1` + `self.linkedbypetsc = 0`). Another alternative is to use `--with-packages-download-dir`, does that work with Spack @balay?

**Reply by lrtfm** [reply]
Adding the option `--with-packages-download-dir=/path/to/download` to the package.py file of `petsc` in spack could work, from my understanding.

**Reply by prj-** [reply]
If either you, @balay, or @joseeroman can give me some guidance on how to use "custom" options in `package.py` for both PETSc and SLEPc, I'd be happy to add the necessary plumbing. It's kind of a must-have option, since many clusters don't have Internet access.

The other route of `--with-hpddm-dir=${HPDDM_DIR}` is again touchy, because not only you need to include `${HPDDM_DIR}/include` for PETSc and SLEPc compilation, but you also need to compile `${HPDDM_DIR}/interface/hpddm_petsc.cpp` after PETSc and SLEPc have been compiled. That's the glue that holds PETSc, SLEPc, and HPDDM together. Since it's non-standard, that would require some added magic in `package.py` of HPDDM. Is there a specific Spack directory where it's allowed to copy sources (and not includes or libraries?).

**Reply by lrtfm** [reply]
In spack, we can copy the directory or file to the install directory:

```
@run_after("install")
def install_source(self):
    prefix = self.spec.prefix
    cp = which("cp")
    cp("-r", os.path.join(self.stage.source_path, "interface"), prefix)              # just copy interface
    cp("-r", os.path.join(self.stage.source_path), os.path.join(prefix, "src"))      # copy all the source code
```
The results looks like this:
```
$ tree -L 1 hpddm-2.2.3-mvgarj7bprlyhumlblnghfdywsyn2rn6
hpddm-2.2.3-mvgarj7bprlyhumlblnghfdywsyn2rn6
├── include
├── interface
└── src
```

**Reply by prj-** [reply]
Yes, we can, but would this be accepted to copy something else than include files after install? This is kind of a nasty hack, so I would understand if this gets rejected. (and in any case, I do believe we really should have a working solution for PETSc option `--with-packages-download-dir`!)

**Reply by joseeroman** [reply]
I don't know enough of spack to advise. But if you can place the tarball somewhere, then you could modify the SLEPc `package.py` to do `--download-hpddm=/path/to/tarball`.

**Reply by prj-** [reply]
Right, but this requires users to touch the engine, and I think we should find a solution where they are only responsible of the steering wheel.

**Reply by prj-** [reply]
This also applies to almost all packages of SLEPc `package.py`.

**Reply by lrtfm** [reply]
> Yes, we can, but would this be accepted to copy something else than include files after install? This is kind of a nasty hack, so I would understand if this gets rejected.

Yes, it is true that copying something else other than include files after installation is not an ideal solution and could be considered a hack. However, in this particular case, I think it is a reasonable approach. 

> But if you can place the tarball somewhere, then you could modify the SLEPc `package.py` to do `--download-hpddm=/path/to/tarball`.

Could we just copy the tarball of hpddm to some directory under the installation of hpddm in spack?

**Reply by balay** [reply]
The spack way to add in hpddm sources(to petsc)  is perhaps something similar to:

https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/trilinos/package.py#L174

**Reply by balay** [reply]
Untested patch [perhaps needs updates wrt versions, locations etc..]
```diff
diff --git a/var/spack/repos/builtin/packages/petsc/package.py b/var/spack/repos/builtin/packages/petsc/package.py
index 8708f12b26..73d7a7b5a6 100644
--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -143,6 +143,14 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     variant("kokkos", default=False, description="Activates support for kokkos and kokkos-kernels")
     variant("fortran", default=True, description="Activates fortran support")
 
+    resource(
+        name="hpddm",
+        git="https://github.com/hpddm/hpddm.git",
+        commit="62414514692d25e0e6785ae806a34974fd7e7183",  # main apr-07-2023
+        placement="arch-spackbuild/externalpackages/hpddm",
+        when="+hpddm @3.19",
+    )
+
     # 3.8.0 has a build issue with MKL - so list this conflict explicitly
     conflicts("^intel-mkl", when="@3.8.0")
 
@@ -367,6 +375,7 @@ def mpi_dependent_options(self):
     def configure_options(self):
         spec = self.spec
         options = [
+            "PETSC_ARCH=arch-spackbuild",
             "--with-ssl=0",
             "--download-c2html=0",
             "--download-sowing=0",
```

**Reply by prj-** [reply]
`+        when="+hpdm @3.19",` -> `+        when="+hpddm @3.19",`

**Reply by balay** [reply]
Thanks - updated the patch above

**Reply by prj-** [reply]
Thank you! @lrtfm, could you please let us know if it works? I'm still try to think of an easier way to do this. For example, I see that a couple of PETSc packages are copying source files in `${PETSC_ARCH}/share`. I don't know if PETSc BuildSystem is using `${PKG_DIR}/share` when being supplied a `--with-pkg-dir=${PKG_DIR}`, but I'm working on this.

**Reply by lrtfm** [reply]
It works as expected. In order to make it works for version `main`, we need 
`+        when="+hpddm @3.19",` -> `+        when="+hpddm @3.19:",`

**Reply by prj-** [reply]
Thanks for the update. I'll try to come up with something to handle `--with-hpddm-dir` and I'll keep you posted.

**Reply by balay** [reply]
each version dependency needs a separate entry in this approach. 

Assuming all hpddm commits (referred to by different version of petsc configure)  are in `main` branch - the following might work.
```diff
diff --git a/var/spack/repos/builtin/packages/petsc/package.py b/var/spack/repos/builtin/packages/petsc/package.py
index 8708f12b26..815cb67a9c 100644
--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -143,6 +143,14 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     variant("kokkos", default=False, description="Activates support for kokkos and kokkos-kernels")
     variant("fortran", default=True, description="Activates fortran support")
 
+    resource(
+        name="hpddm",
+        git="https://github.com/hpddm/hpddm.git",
+        branch="main",
+        placement="arch-spackbuild/externalpackages/git.hpddm",
+        when="+hpddm",
+    )
+
     # 3.8.0 has a build issue with MKL - so list this conflict explicitly
     conflicts("^intel-mkl", when="@3.8.0")
 
@@ -367,6 +375,7 @@ def mpi_dependent_options(self):
     def configure_options(self):
         spec = self.spec
         options = [
+            "PETSC_ARCH=arch-spackbuild",
             "--with-ssl=0",
             "--download-c2html=0",
             "--download-sowing=0",
```

Well the caveat here is - it might fail with shallow clones - but when I checked the hpddm clone - eventhough it was shallow - it appeared to have all the commits in this brach

**Reply by prj-** [reply]
Ah, yes, that's much neater.
